File: android/app/src/main/java/com/flow/app/Federation.kt
package com.flow.app

import android.util.Base64
import android.util.Log
import com.google.crypto.tink.Aead
import com.google.crypto.tink.subtle.XChaCha20Poly1305
import kotlinx.serialization.json.JsonObject
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import java.security.MessageDigest
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Federation peer (spec §4) — Android side.
 *
 * Each node is a symmetric peer. On Android we implement the **client** half with
 * OkHttp's WebSocket: after scanning the laptop's QR we dial ws://ip:port and run
 * the HELLO/QUERY/RESULTS/FETCH/SURFACE protocol from shared/protocol.md.
 *
 * NOTE on the "server-ish" half: OkHttp has no embedded WS server. For the hackathon
 * the laptop is the pairing initiator and accepts the connection, so the phone only
 * needs the client. If you want the phone to also accept inbound peers (true mesh),
 * add Ktor's embedded server behind the same [PeerLink] interface — see the TODO in
 * [FederationManager]. The desktop engine listens on Config.WS_PORT.
 *
 * Every frame is AEAD-encrypted with a key derived from the pairing PSK (spec §0.5).
 * shared/config.json specifies XChaCha20-Poly1305 + HKDF-SHA256. The JDK/Android
 * crypto providers do not ship XChaCha20-Poly1305, so this skeleton uses a CLEARLY
 * DOCUMENTED interop-friendly stub ([AeadChannel]) built on AES-256-GCM + HKDF-SHA256.
 *
 *   >>> Before interop with the desktop engine, replace AeadChannel with a real
 *       XChaCha20-Poly1305 implementation (Tink's `XChaCha20Poly1305` /
 *       libsodium via JNI) so both ends agree on the AEAD. The wrap/unwrap framing
 *       (nonce-prefixed, base64) is kept identical so only the cipher swaps out.
 */

private const val TAG = "Flow/Federation"

/** Abstraction over a single peer connection so the transport can be swapped. */
interface PeerLink {
    val deviceId: String
    fun send(plaintextJson: String)
    fun close()
}

/**
 * AEAD channel keyed by the pairing PSK.
 *
 * Wire framing (must match the desktop engine): base64( nonce(12B) || ciphertext||tag ).
 * Key derivation: HKDF-SHA256(psk, salt="flow-fed-v1", info="aead-key") -> 32 bytes.
 *
 * Cipher: **AES-256-GCM stub** standing in for the spec's XChaCha20-Poly1305 — see file
 * header. The nonce is a 12-byte monotonic counter prefixed with 4 random bytes; a
 * replay window rejects out-of-order/duplicate counters (spec §0.5).
 */
class AeadChannel(psk: ByteArray) {
    // Session key + cipher chosen to interoperate byte-for-byte with the desktop
    // engine's flow/crypto.py SecureChannel:
    //   key    = HKDF-SHA256(ikm=psk, salt=32 zero bytes, info="flow/session/v1") -> 32B
    //   cipher = XChaCha20-Poly1305 (IETF, 24-byte random nonce) via Tink
    //   frame  = base64( nonce(24) || ciphertext||tag )
    // (Previously an AES-256-GCM stub with a 12-byte counter nonce — incompatible
    //  with the desktop, so cross-device frames never decrypted.)
    private val aead: Aead = XChaCha20Poly1305(
        hkdfSha256(psk, ByteArray(32), "flow/session/v1".toByteArray(Charsets.UTF_8), 32)
    )

    fun seal(plaintext: String): String {
        val frame = aead.encrypt(plaintext.toByteArray(Charsets.UTF_8), EMPTY_AAD)
        return Base64.encodeToString(frame, Base64.NO_WRAP)
    }

    /** @return plaintext JSON, or null if the frame is malformed or fails authentication. */
    fun open(frameB64: String): String? {
        return try {
            val frame = Base64.decode(frameB64, Base64.NO_WRAP)
            String(aead.decrypt(frame, EMPTY_AAD), Charsets.UTF_8)
        } catch (t: Throwable) {
            Log.w(TAG, "open() failed: ${t.message}")
            null
        }
    }

    companion object {
        // XChaCha20-Poly1305 with empty associated data, matching the desktop's aad=None.
        private val EMPTY_AAD = ByteArray(0)

        /** HKDF-SHA256 (extract + expand) — matches the desktop engine's KDF. */
        fun hkdfSha256(ikm: ByteArray, salt: ByteArray, info: ByteArray, length: Int): ByteArray {
            val mac = Mac.getInstance("HmacSHA256")
            // extract
            mac.init(SecretKeySpec(salt, "HmacSHA256"))
            val prk = mac.doFinal(ikm)
            // expand
            mac.init(SecretKeySpec(prk, "HmacSHA256"))
            val out = ByteArray(length)
            var t = ByteArray(0)
            var pos = 0
            var counter = 1
            while (pos < length) {
                mac.reset()
                mac.update(t)
                mac.update(info)
                mac.update(counter.toByte())
                t = mac.doFinal()
                val n = minOf(t.size, length - pos)
                System.arraycopy(t, 0, out, pos, n)
                pos += n
                counter++
            }
            return out
        }
    }
}

/**
 * Parses and validates the pairing QR payload (spec §Pairing). Refuses to pair on a
 * major protocol-version mismatch.
 */
object Pairing {
    fun parse(qrText: String): Result<PairingInfo> = runCatching {
        val info = FlowJson.decodeFromString(PairingInfo.serializer(), qrText)
        require(info.v == Config.PROTOCOL_VERSION) {
            "protocol version mismatch: peer=${info.v} self=${Config.PROTOCOL_VERSION}"
        }
        val psk = Base64.decode(info.psk, Base64.DEFAULT)
        require(psk.size == Config.Crypto.PSK_BYTES) { "psk must be ${Config.Crypto.PSK_BYTES} bytes" }
        info
    }

    fun pskBytes(info: PairingInfo): ByteArray = Base64.decode(info.psk, Base64.DEFAULT)
}

/**
 * Owns the OkHttp client, the encrypted channel, and the protocol handlers.
 * One instance per process (held by [IndexingService]); connect once per paired peer.
 */
class FederationManager(
    private val selfDeviceId: String,
    private val selfName: String,
    private val index: Index,
    private val caps: Caps
) {
    private val client = OkHttpClient.Builder()
        .pingInterval(15, TimeUnit.SECONDS)       // heartbeat (spec §4.2)
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .build()

    private val peers = ConcurrentHashMap<String, OkHttpPeer>()

    /** Pending QUERY correlation: query_id -> collector. */
    private val pending = ConcurrentHashMap<String, QueryCollector>()

    /** UI/Service hook for proactive recall (SURFACE) and incoming results. */
    var onSurface: ((Surface) -> Unit)? = null

    /**
     * Dial a peer described by a scanned QR. Auto-reconnect is left as a TODO hook
     * (re-invoke connect() on onFailure with backoff).
     */
    fun connect(info: PairingInfo) {
        val channel = AeadChannel(Pairing.pskBytes(info))
        val url = "ws://${info.ip}:${info.port}/flow"
        val request = Request.Builder().url(url).build()
        val peer = OkHttpPeer(channel)
        val ws = client.newWebSocket(request, PeerListener(peer))
        peer.attach(ws)
        Log.i(TAG, "dialing $url")
    }

    /**
     * Federated ask (spec §5.1 client side): broadcast QUERY to all peers and collect
     * RESULTS within the timeout. Local search is done by the caller and fused via [Fusion].
     */
    fun broadcastQuery(text: String, topK: Int = Config.TOP_K): String {
        val queryId = UUID.randomUUID().toString()
        // Snapshot the fan-out target count so the asker can stop polling as soon as every
        // peer it broadcast to has replied (spec §4.4: partial allowed, return early).
        val targets = peers.values.toList()
        pending[queryId] = QueryCollector(expectedPeers = targets.size)
        val q = FlowJson.encodeToString(Query.serializer(), Query(query_id = queryId, text = text, top_k = topK))
        targets.forEach { it.send(q) }
        return queryId
    }

    /** Block-free collection: returns whatever RESULTS arrived for [queryId] so far. */
    fun collectedHits(queryId: String): List<Hit> = pending[queryId]?.hits().orEmpty()

    /** True once every peer this query was broadcast to has returned its RESULTS. */
    fun allPeersReplied(queryId: String): Boolean = pending[queryId]?.isComplete() ?: true

    fun finishQuery(queryId: String) { pending.remove(queryId) }

    private inner class PeerListener(private val peer: OkHttpPeer) : WebSocketListener() {
        override fun onOpen(webSocket: WebSocket, response: Response) {
            // HELLO handshake.
            val hello = Hello(device_id = selfDeviceId, name = selfName, caps = caps)
            peer.send(FlowJson.encodeToString(Hello.serializer(), hello))
        }

        override fun onMessage(webSocket: WebSocket, text: String) {
            val plaintext = peer.channel.open(text) ?: return   // decrypt; drop bad frames
            handle(plaintext, peer)
        }

        override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
            Log.w(TAG, "peer ${peer.deviceId} failed: ${t.message}")
            peers.remove(peer.deviceId)
            // TODO: reconnect with exponential backoff (spec §4.2 "survive a Wi-Fi blip").
        }

        override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
            peers.remove(peer.deviceId)
        }
    }

    /** Dispatch an inbound plaintext message by its `type`. */
    private fun handle(plaintext: String, peer: OkHttpPeer) {
        when (Wire.peekType(plaintext)) {
            MsgType.HELLO -> {
                val hello = FlowJson.decodeFromString(Hello.serializer(), plaintext)
                peer.deviceId = hello.device_id
                peers[hello.device_id] = peer
                Log.i(TAG, "paired with ${hello.name} (${hello.device_id})")
            }
            MsgType.QUERY -> {
                // A peer asked us — search OUR index only and reply with hits (snippets+thumbs).
                val q = FlowJson.decodeFromString(Query.serializer(), plaintext)
                val qv = StubEmbedHook.embed?.invoke(q.text) ?: return
                val hits = index.searchText(qv, q.top_k)
                val res = Results(query_id = q.query_id, device_id = selfDeviceId, hits = hits)
                peer.send(FlowJson.encodeToString(Results.serializer(), res))
            }
            MsgType.RESULTS -> {
                val r = FlowJson.decodeFromString(Results.serializer(), plaintext)
                pending[r.query_id]?.add(r.hits)   // also counts this peer as having replied
            }
            MsgType.FETCH -> {
                val f = FlowJson.decodeFromString(Fetch.serializer(), plaintext)
                val item = index.get(f.item_id)
                // TODO: read the real file at item.file_ref and base64 it. Thumb fallback for now.
                val blob = item?.thumb_b64 ?: ""
                val fr = FetchResult(item_id = f.item_id, mime = "image/jpeg", blob_b64 = blob)
                peer.send(FlowJson.encodeToString(FetchResult.serializer(), fr))
            }
            MsgType.FETCH_RESULT -> {
                // Hydrated lazily on open; the UI layer subscribes via its own callback. (TODO)
            }
            MsgType.SURFACE -> {
                onSurface?.invoke(FlowJson.decodeFromString(Surface.serializer(), plaintext))
            }
            else -> Log.w(TAG, "unknown message type")
        }
    }

    fun shutdown() {
        peers.values.forEach { it.close() }
        peers.clear()
        client.dispatcher.executorService.shutdown()
    }

    /** Concrete OkHttp-backed peer link; seals every outbound frame. */
    private class OkHttpPeer(val channel: AeadChannel) : PeerLink {
        override var deviceId: String = "pending"
        private var ws: WebSocket? = null
        fun attach(socket: WebSocket) { ws = socket }
        override fun send(plaintextJson: String) { ws?.send(channel.seal(plaintextJson)) }
        override fun close() { ws?.close(1000, "bye") }
    }
}

/**
 * Bridge so [FederationManager] can embed a peer's QUERY text without taking a hard
 * dependency on [Inference]. [IndexingService] sets this to `inference::embedText`.
 */
object StubEmbedHook {
    @Volatile var embed: ((String) -> FloatArray)? = null
}

/**
 * Accumulates RESULTS hits for one query_id (partial results allowed; spec §4.4).
 *
 * Tracks how many of the [expectedPeers] have replied so the asker can stop waiting
 * early once everyone has answered instead of always blocking for the full timeout.
 */
private class QueryCollector(private val expectedPeers: Int) {
    private val all = ArrayList<Hit>()
    private var replied = 0
    @Synchronized fun add(hits: List<Hit>) { all.addAll(hits); replied++ }
    @Synchronized fun hits(): List<Hit> = ArrayList(all)
    /** True once every broadcast target has replied (vacuously true when there are no peers). */
    @Synchronized fun isComplete(): Boolean = replied >= expectedPeers
}

/** Stable per-install device id derived from a random seed; persisted by the caller. */
fun deriveDeviceId(seed: String): String {
    val md = MessageDigest.getInstance("SHA-256")
    return Base64.encodeToString(md.digest(seed.toByteArray()), Base64.NO_WRAP or Base64.URL_SAFE).take(16)
}

File: desktop/flow/federation.py
"""Federation — peer mesh over encrypted WebSockets + QR pairing (§4).

Each node runs both a WebSocket **server** and **client** (symmetric mesh).
All frames are AEAD-encrypted via :class:`flow.crypto.SecureChannel`; the
plaintext inside is the JSON of a protocol message.

Responsibilities:
  * QR pairing: produce / decode the ``{ip, port, psk, v}`` JSON (protocol.md).
  * Handle inbound QUERY (search local index, reply RESULTS), FETCH/FETCH_RESULT,
    HELLO, SURFACE.
  * Fan-out a QUERY to all peers and collect RESULTS within ``query_timeout_ms``
    (partial results allowed).

The federation layer is transport + routing only; the actual search and answer
composition live in ``brain``. A ``search_fn`` callback is injected to avoid a
circular import.
"""

from __future__ import annotations

import asyncio
import base64
import json
import socket
import time
from dataclasses import dataclass
from typing import Awaitable, Callable, Dict, List, Optional

from . import config
from .crypto import SecureChannel, ReplayError, random_psk
from .protocol import (
    Fetch, FetchResult, Hello, Query, Results, Surface,
    decode_message, FETCH, FETCH_RESULT, HELLO, QUERY, RESULTS, SURFACE,
    new_id, Hit,
)
from .telemetry import TELEMETRY

try:
    import websockets  # type: ignore
    _WS_AVAILABLE = True
except Exception:  # pragma: no cover
    websockets = None  # type: ignore
    _WS_AVAILABLE = False


# Callback signatures injected by brain/server.
SearchFn = Callable[[str, int], List[Hit]]            # (query_text, top_k) -> local hits
FetchFn = Callable[[str], Optional["FetchResult"]]    # (item_id) -> FetchResult or None
SurfaceFn = Callable[[Surface], None]                 # inbound proactive push


# ---------------------------------------------------------------------------
# QR pairing payload
# ---------------------------------------------------------------------------
def local_ip() -> str:
    """Best-effort LAN IP (the address a peer would dial)."""
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        s.connect(("8.8.8.8", 80))
        ip = s.getsockname()[0]
        s.close()
        return ip
    except Exception:
        return "127.0.0.1"


def make_pairing_payload(psk: Optional[bytes] = None,
                         ip: Optional[str] = None,
                         port: int = config.WS_PORT) -> Dict[str, object]:
    """Build the QR JSON: ``{ip, port, psk(base64 32B), v}``."""
    psk = psk or random_psk()
    return {
        "ip": ip or local_ip(),
        "port": int(port),
        "psk": base64.b64encode(psk).decode("ascii"),
        "v": config.PROTOCOL_VERSION,
    }


def parse_pairing_payload(data) -> Dict[str, object]:
    """Decode the QR JSON (str or dict) into a normalized dict with raw psk bytes."""
    if isinstance(data, (bytes, str)):
        obj = json.loads(data)
    else:
        obj = dict(data)
    if int(obj.get("v", config.PROTOCOL_VERSION)) != config.PROTOCOL_VERSION:
        raise ValueError("pairing protocol version mismatch")
    psk = base64.b64decode(obj["psk"])
    if len(psk) != config.PSK_BYTES:
        raise ValueError("bad psk length")
    return {"ip": obj["ip"], "port": int(obj["port"]), "psk": psk,
            "v": int(obj.get("v", config.PROTOCOL_VERSION))}


def make_qr_png(payload: Optional[Dict[str, object]] = None) -> bytes:
    """Render the pairing payload as a PNG (bytes)."""
    import io
    import qrcode

    payload = payload or make_pairing_payload()
    img = qrcode.make(json.dumps(payload))
    buf = io.BytesIO()
    img.save(buf, format="PNG")
    return buf.getvalue()


# ---------------------------------------------------------------------------
# Peer connections
# ---------------------------------------------------------------------------
@dataclass
class Peer:
    device_id: str
    name: str
    ws: object  # websocket connection
    channel: SecureChannel
    caps: Dict[str, object]


class Federation:
    """Symmetric peer-mesh node: WS server + client, with one shared PSK.

    For the hackathon demo a single PSK is shared across the mesh (the QR-paired
    key). Each connection gets its own :class:`SecureChannel` instance so replay
    windows are per-connection.
    """

    def __init__(self, device_id: str, name: str,
                 psk: Optional[bytes] = None,
                 search_fn: Optional[SearchFn] = None,
                 fetch_fn: Optional[FetchFn] = None,
                 surface_fn: Optional[SurfaceFn] = None,
                 port: int = config.WS_PORT) -> None:
        self.device_id = device_id
        self.name = name
        self.port = port
        self.psk = psk or random_psk()
        self.search_fn = search_fn
        self.fetch_fn = fetch_fn
        self.surface_fn = surface_fn

        self._peers: Dict[str, Peer] = {}
        self._server = None
        # query_id -> {device_id: Results}
        self._pending: Dict[str, Dict[str, Results]] = {}
        self._pending_events: Dict[str, asyncio.Event] = {}
        # item_id correlation for FETCH responses
        self._fetch_waiters: Dict[str, asyncio.Future] = {}

    # ---- pairing helpers ----
    def pairing_payload(self) -> Dict[str, object]:
        return make_pairing_payload(self.psk, port=self.port)

    def set_psk(self, psk: bytes) -> None:
        self.psk = psk

    # ---- server ----
    async def start_server(self) -> None:
        if not _WS_AVAILABLE:
            raise RuntimeError("websockets not installed; cannot start server")
        self._server = await websockets.serve(self._on_connection, "0.0.0.0", self.port)
        TELEMETRY.record("federation_server", "n/a", 0.0,
                         extra={"port": self.port, "device_id": self.device_id})

    async def _on_connection(self, ws) -> None:
        channel = SecureChannel(self.psk)
        peer_id: Optional[str] = None
        try:
            async for raw in ws:
                msg = self._decrypt(channel, raw)
                if msg is None:
                    continue
                peer_id = await self._dispatch(ws, channel, msg, peer_id)
        except Exception:
            pass
        finally:
            if peer_id and peer_id in self._peers:
                del self._peers[peer_id]

    # ---- client ----
    async def connect(self, ip: str, port: int, psk: Optional[bytes] = None) -> Peer:
        if not _WS_AVAILABLE:
            raise RuntimeError("websockets not installed; cannot connect")
        psk = psk or self.psk
        self.psk = psk  # adopt the paired key for the mesh
        channel = SecureChannel(psk)
        uri = f"ws://{ip}:{port}"
        ws = await websockets.connect(uri, max_size=16 * 1024 * 1024)

        # Handshake: send HELLO.
        hello = Hello(device_id=self.device_id, name=self.name,
                      caps={"has_llm": True, "tops": 0, "battery": None})
        await ws.send(channel.seal_b64(json.dumps(hello.to_dict()).encode()))

        peer = Peer(device_id=f"pending:{ip}:{port}", name="", ws=ws,
                    channel=channel, caps={})
        self._peers[peer.device_id] = peer
        # Pump inbound frames from this connection in the background.
        asyncio.ensure_future(self._client_pump(peer))
        return peer

    async def _client_pump(self, peer: Peer) -> None:
        try:
            async for raw in peer.ws:
                msg = self._decrypt(peer.channel, raw)
                if msg is None:
                    continue
                await self._dispatch(peer.ws, peer.channel, msg, peer.device_id)
        except Exception:
            pass

    # ---- crypto plumbing ----
    def _decrypt(self, channel: SecureChannel, raw):
        try:
            if isinstance(raw, (bytes, bytearray)):
                pt = channel.open(bytes(raw))
            else:
                pt = channel.open_b64(raw)
        except ReplayError:
            TELEMETRY.record("replay_rejected", "n/a", 0.0)
            return None
        except Exception:
            TELEMETRY.record("frame_rejected", "n/a", 0.0)
            return None
        try:
            return decode_message(json.loads(pt.decode("utf-8")))
        except Exception:
            return None

    async def _send(self, ws, channel: SecureChannel, msg_dict: dict) -> None:
        await ws.send(channel.seal_b64(json.dumps(msg_dict).encode()))

    # ---- dispatch ----
    async def _dispatch(self, ws, channel: SecureChannel, msg, peer_id):
        if isinstance(msg, Hello):
            self._peers[msg.device_id] = Peer(
                device_id=msg.device_id, name=msg.name, ws=ws,
                channel=channel, caps=msg.caps,
            )
            return msg.device_id

        if isinstance(msg, Query):
            hits = self.search_fn(msg.text, msg.top_k) if self.search_fn else []
            res = Results(query_id=msg.query_id, device_id=self.device_id, hits=hits)
            await self._send(ws, channel, res.to_dict())
            return peer_id

        if isinstance(msg, Results):
            bucket = self._pending.get(msg.query_id)
            if bucket is not None:
                bucket[msg.device_id] = msg
                ev = self._pending_events.get(msg.query_id)
                if ev is not None:
                    ev.set()
            return peer_id

        if isinstance(msg, Fetch):
            fr = self.fetch_fn(msg.item_id) if self.fetch_fn else None
            if fr is None:
                fr = FetchResult(item_id=msg.item_id, mime="application/octet-stream",
                                 blob_b64="")
            await self._send(ws, channel, fr.to_dict())
            return peer_id

        if isinstance(msg, FetchResult):
            fut = self._fetch_waiters.get(msg.item_id)
            if fut is not None and not fut.done():
                fut.set_result(msg)
            return peer_id

        if isinstance(msg, Surface):
            if self.surface_fn:
                self.surface_fn(msg)
            return peer_id

        return peer_id

    # ---- fan-out QUERY ----
    async def query_peers(self, text: str, top_k: int = config.TOP_K,
                          timeout_ms: int = config.QUERY_TIMEOUT_MS) -> List[Results]:
        """Broadcast a QUERY to all peers; collect RESULTS within timeout.

        Returns whatever arrived before the deadline (partial allowed).
        """
        if not self._peers:
            return []
        query_id = new_id()
        self._pending[query_id] = {}
        self._pending_events[query_id] = asyncio.Event()
        q = Query(query_id=query_id, text=text, top_k=top_k)

        targets = list(self._peers.values())
        for peer in targets:
            try:
                await self._send(peer.ws, peer.channel, q.to_dict())
            except Exception:
                continue

        deadline = time.perf_counter() + (timeout_ms / 1000.0)
        try:
            while time.perf_counter() < deadline:
                bucket = self._pending[query_id]
                if len(bucket) >= len(targets):
                    break
                remaining = deadline - time.perf_counter()
                if remaining <= 0:
                    break
                ev = self._pending_events[query_id]
                try:
                    await asyncio.wait_for(ev.wait(), timeout=remaining)
                    ev.clear()
                except asyncio.TimeoutError:
                    break
            return list(self._pending[query_id].values())
        finally:
            self._pending.pop(query_id, None)
            self._pending_events.pop(query_id, None)

    # ---- fetch from a peer ----
    async def fetch_from_peer(self, device_id: str, item_id: str,
                              timeout_ms: int = 5000) -> Optional[FetchResult]:
        peer = self._peers.get(device_id)
        if peer is None:
            return None
        fut: asyncio.Future = asyncio.get_event_loop().create_future()
        self._fetch_waiters[item_id] = fut
        try:
            await self._send(peer.ws, peer.channel, Fetch(item_id=item_id).to_dict())
            return await asyncio.wait_for(fut, timeout=timeout_ms / 1000.0)
        except asyncio.TimeoutError:
            return None
        finally:
            self._fetch_waiters.pop(item_id, None)

    async def broadcast_surface(self, surface: Surface) -> None:
        for peer in list(self._peers.values()):
            try:
                await self._send(peer.ws, peer.channel, surface.to_dict())
            except Exception:
                continue

    def peer_list(self) -> List[Dict[str, object]]:
        return [{"device_id": p.device_id, "name": p.name, "caps": p.caps}
                for p in self._peers.values()]

    async def close(self) -> None:
        for p in list(self._peers.values()):
            try:
                await p.ws.close()
            except Exception:
                pass
        if self._server is not None:
            self._server.close()
            await self._server.wait_closed()
